controllers/workspace/metrics/update.go

```diff
@@ -22,12 +22,16 @@ import (
 
 	"github.com/devfile/devworkspace-operator/pkg/conditions"
 	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
 )
 
 // WorkspaceStarted updates metrics for workspaces entering the 'Starting' phase, given a workspace. If an error is
 // encountered, the provided logger is used to log the error.
 func WorkspaceStarted(wksp *dw.DevWorkspace, log logr.Logger) {
-	incrementMetricForWorkspace(workspaceTotal, wksp, log)
+	_, ok := wksp.GetAnnotations()[constants.DevWorkspaceStartedAtAnnotation]
+	if !ok {
+		incrementMetricForWorkspace(workspaceTotal, wksp, log)
+	}
```

Comment:
I think this approach is good, but will only solve half the problem. With this PR, we
1. Set the started-at annotation on the DevWorkspace on start
2. Only increment the started metric if the started-at annotation is not present

This kind of reveals how weirdly we've implemented startup/running metrics :smile: -- `WorkspaceStarted` is called in two places (AFAIK): 
* In a guarded conditional that only executes if the workspace's phase is transitioning from `Stopped` or `Failed/Failing` [here](https://github.com/devfile/devworkspace-operator/blob/9f867aeac1d014d315f53594953e7a4c032a9943/controllers/workspace/devworkspace_controller.go#L186)
* Whenever the Workspace switches to `Starting` from another phase [here](https://github.com/devfile/devworkspace-operator/blob/9f867aeac1d014d315f53594953e7a4c032a9943/controllers/workspace/status.go#L241-L242)

The second case is the one we're limiting with the `started-at` annotation, but _whenever_ the second case is called, the workspace will have the `started-at` annotation -- we may as well just remove the Starting condition from `updateMetricsForPhase`.

One thing that's missing from this PR is that we call [`WorkspaceRunning`](https://github.com/devfile/devworkspace-operator/blob/9f867aeac1d014d315f53594953e7a4c032a9943/controllers/workspace/status.go#L237-L238) every time the workspace goes from `Starting -> Running`, even if that workspace restarted due to e.g. a crashed container. This can result in e.g. `1` workspace Start and `n` workspace Running metrics for a single workspace.

A DevWorkspace that should (:crossed_fingers:) reproduce this is something like
```yaml
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: sleep-ten
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:latest
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "sleep"
           - "10s"
```

Reply:
I think the new commit should fix this. With the new commit the process is essentially:

- Workspace started metric is only added on [line 186](https://github.com/devfile/devworkspace-operator/pull/686/files?diff=unified&w=0#diff-05e72230161a4d41c5a27fb453041c78839c193a9ddcb5066b1cb62c5a399529R186)
- If the workspace eventually is running, [line 209](https://github.com/devfile/devworkspace-operator/pull/686/files?diff=unified&w=0#diff-05e72230161a4d41c5a27fb453041c78839c193a9ddcb5066b1cb62c5a399529R209) records that the workspace is running (as long as it doesn't have a startedAt annotation yet) then the startAt time is synced to the cluster
- If the workspace is then running -> restarted -> running -> restarted
    - running won't increase because the startedAt annotation is already set the first time it got to running
- If the workspace crash loops because of something like ImagePullBackOff
    - starting gets updated but running doesn't  
- If the workspace is scaled down
    - startedAt is removed
    - restarting it again will trigger line 186 and then repeat all the other actions